lib/Basics/VelocyPackHelper.cpp

```diff
@@ -387,6 +391,188 @@ int VelocyPackHelper::compareNumberValues(VPackValueType lhsType,
   return (l < r ? -1 : 1);
 }
 
+template<typename T>
+int comp(T a, T b) {
+  if (a == b) {
+    return 0;
+  }
+  return (a < b ? -1 : 1);
+}
+
+static int compareInt64UInt64(int64_t i, uint64_t u) {
+  if (i < 0) {
+    return -1;
+  }
+  return comp<uint64_t>(static_cast<uint64_t>(i), u);
+}
+
+constexpr uint64_t uint64_2_52 = uint64_t{1} << 52;
+constexpr uint64_t uint64_2_63 = uint64_t{1} << 63;
+constexpr double double_2_52 = static_cast<double>(uint64_2_52);
+const double double_2_64 = ldexp(1.0, 64);
+
+static int compareUInt64Double(uint64_t u, double d) {
+  if (std::isnan(d)) {
+    return -1;
```

Comment:
doesn't this mean that a NaN double value is considered _higher_ than any other integer value? _higher_ here because the double value is the 2nd parameter of the function, and we are returning -1 here. IMHO a NaN double value should rather compare lower than any integer value.

Reply:
This is up for debate. Since all actual comparisons (<, <=, >, >=, ==) involving a NaN will always return `false`, we need to do something, or else we run the risk that our order on the set of all VelocyPack values is not transitive. Therefore, we have to make a choice. Either we make NaN smaller than all numbers or larger than all numbers. For now, I have chosen larger than all numbers, but there is no good reason either way, as far as I can tell.

Reply:
There is no good way out here, for the following reason: in the legacy comparator, all comparisons with NaN are `false`. So, in existing indexes (provided the user has inserted a NaN), the question as to where the item is placed in the sorted index depended on whether the NaN was on the left hand side of the comparison or on the right hand side, which is utterly random. Therefore, we have to live with the fact that in existing indexes, a NaN entry might be anywhere in the list. The new comparator will now sort it behind all numbers (or - if we choose so - before all numbers), which is almost certainly different from what has happened so far. So people with NaN in their indexed columns are pretty likely to run into difficulties and will not be a very easy migration case.
On the other hand, if people had inserted more NaN in their sorted indexes, we would have stumbled over this much earlier.